src/detect.c

```diff
@@ -1486,6 +1507,8 @@ static void DetectRunTx(ThreadVars *tv,
                     tx.tx_ptr, tx.tx_id,
                     flow_flags & STREAM_TOSERVER ? "toserver" : "toclient",
                     new_detect_flags);
+        if (alproto == ALPROTO_HTTP)
```

Comment:
I would prefer a solution where we can avoid adding checks here.

One way would be to make sure `HttpServerBodyGetDataCallback` is still called only once, but we cache the output for the duration of the current run. `DetectEngineInspectBufferHttpHeader` does something similar for HTTP headers.

Reply:
Though I have to say I don't really like the solution `http_header` uses either. But it does put the cost where it belongs, in the keyword implementation. This way its not a cost that all the other keywords and protocols also have.

Reply:
@victorjulien This implements your second suggestion -- please review when you have time.